# General EDA of our merged Data

In [48]:
import pandas as pd
import numpy as np
import dataprep as dp
from dataprep.eda import plot
import os

# Load in our data
filepath = '../../Data/entitiesResolved/merged_data.csv'
data = pd.read_csv(filepath)

data.head(-5)

/var/folders/jd/f_rvzywd5j7g5f9ckcp7wsxr0000gp/T/ipykernel_35480/3990006058.py:9: DtypeWarning: Columns (37,40,49,67) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv(filepath)


POSITION            PLAYER    TEAM_x          TOI  GP     TOI/GP  \
0            r       adam burish       chi   951.616667  81  11.748354   
1            d        adam foote  cbj, col  1754.250000  75  23.390000   
2            r         adam hall       pit   546.150000  46  11.872826   
3            r       ales hemsky       edm  1374.483333  74  18.574099   
4            r      ales kotalik       buf  1212.050000  79  15.342405   
...        ...               ...       ...          ...  ..        ...   
12947        c  zach aston-reese       det     6.350000   1   6.350000   
12948        l       zach benson       buf   876.383333  60  14.606389   
12949        d     zach bogosian  min, t.b   964.033333  55  17.527879   
12950        c         zach dean       stl     8.683333   1   8.683333   
12951        l        zach hyman       edm  1292.750000  66  19.587121   

       GOALS/60  TOTAL ASSISTS/60  FIRST ASSISTS/60  SECOND ASSISTS/60  ...  \
0          0.25              0.25              0.19               0.06  ...   
1          0.03              0.51              0.10               0.41  ...   
2          0.22              0.44              0.11               0.33  ...   
3          0.87              2.23              1.57               0.65  ...   
4          1.14              0.99              0.35               0.64  ...   
...         ...               ...               ...                ...  ...   
12947      0.00              0.00              0.00               0.00  ...   
12948      0.62              0.96              0.55               0.41  ...   
12949      0.19              0.56              0.25               0.31  ...   
12950      0.00              0.00              0.00               0.00  ...   
12951      2.23              0.88              0.46               0.42  ...   

       EXP. YEAR     CAP HIT         AAV      SALARY  BASE SALARY  S.BONUS  \
0           2008    $575,000    $575,000    $575,000     $500,000  $75,000   
1           2008  $4,600,000  $4,600,000  $4,600,000   $4,600,000       $0   
2           2008    $525,000    $525,000    $525,000     $525,000       $0   
3           2012  $4,100,000  $4,100,000  $3,600,000   $3,600,000       $0   
4           2009  $2,333,333  $2,333,333  $2,500,000   $2,500,000       $0   
...          ...         ...         ...         ...          ...      ...   
12947       2024    $775,000    $775,000    $775,000     $775,000       $0   
12948       2026    $950,000  $1,600,000    $950,000     $855,000  $95,000   
12949       2024    $850,000    $850,000  $1,050,000   $1,050,000       $0   
12950       2026    $852,500    $863,333    $892,500     $800,000  $92,500   
12951       2028  $5,500,000  $5,500,000  $7,650,000   $7,650,000       $0   

        P.BONUS   SEASON  Y_SALARY_CAP  Y_SALARY_CAP_PERCENTAGE  
0            $0  2007-08      50300000                 0.011431  
1            $0  2007-08      50300000                 0.091451  
2            $0  2007-08      50300000                 0.010437  
3            $0  2007-08      50300000                 0.081511  
4            $0  2007-08      50300000                 0.046388  
...         ...      ...           ...                      ...  
12947        $0  2023-24      83500000                 0.009281  
12948  $650,000  2023-24      83500000                 0.011377  
12949        $0  2023-24      83500000                 0.010180  
12950   $32,500  2023-24      83500000                 0.010210  
12951        $0  2023-24      83500000                 0.065868  

[12952 rows x 119 columns]

In [49]:
data.columns.to_list()

['POSITION',
 'PLAYER',
 'TEAM_x',
 'TOI',
 'GP',
 'TOI/GP',
 'GOALS/60',
 'TOTAL ASSISTS/60',
 'FIRST ASSISTS/60',
 'SECOND ASSISTS/60',
 'TOTAL POINTS/60',
 'IPP',
 'SHOTS/60',
 'SH%',
 'IXG/60',
 'ICF/60',
 'IFF/60',
 'ISCF/60',
 'IHDCF/60',
 'RUSH ATTEMPTS/60',
 'REBOUNDS CREATED/60',
 'PIM/60',
 'TOTAL PENALTIES/60',
 'MINOR/60',
 'MAJOR/60',
 'MISCONDUCT/60',
 'PENALTIES DRAWN/60',
 'GIVEAWAYS/60',
 'TAKEAWAYS/60',
 'HITS/60',
 'HITS TAKEN/60',
 'SHOTS BLOCKED/60',
 'FACEOFFS WON/60',
 'FACEOFFS LOST/60',
 'FACEOFFS %',
 'CF/60',
 'CA/60',
 'CF%',
 'FF/60',
 'FA/60',
 'FF%',
 'SF/60',
 'SA/60',
 'SF%',
 'GF/60',
 'GA/60',
 'GF%',
 'XGF/60',
 'XGA/60',
 'XGF%',
 'SCF/60',
 'SCA/60',
 'SCF%',
 'HDCF/60',
 'HDCA/60',
 'HDCF%',
 'HDGF/60',
 'HDGA/60',
 'HDGF%',
 'MDCF/60',
 'MDCA/60',
 'MDCF%',
 'MDGF/60',
 'MDGA/60',
 'MDGF%',
 'LDCF/60',
 'LDCA/60',
 'LDCF%',
 'LDGF/60',
 'LDGA/60',
 'LDGF%',
 'ON-ICE SH%',
 'ON-ICE SV%',
 'PDO',
 'OFF.\xa0ZONE STARTS/60',
 'NEU.\xa0ZONE STARTS/60'

In [50]:
# Change the name of column 'Y_SALARY_CAP' to 'SALARY_CAP' and 'Y_SALARY_CAP_PERCENTAGE' to 'SALARY_CAP_PERCENTAGE'
data.rename(columns={'Y_SALARY_CAP': 'SALARY CAP'}, inplace=True)
data.rename(columns={'Y_SALARY_CAP_PERCENTAGE': 'SALARY CAP PERCENTAGE'}, inplace=True)

Notice that we have a few similar column names: TEAM and AGE that are still overlapping from our merge. We should decide which of these to keep. 

## First, lets take a look at TEAM

In [51]:
print(data['TEAM_x'].unique())
print(data['TEAM_x'].nunique())
print(data['TEAM_y'].unique())
print(data['TEAM_y'].nunique())

['chi' 'cbj, col' 'pit' 'edm' 'buf' 'wsh' 'cgy' 'van' 'nsh' 'tor' 'atl'
 'det' 'mtl' 'ott' 'bos' 'ana' 'car, chi' 'n.j' 'nyi' 'ana, stl' 'l.a'
 'dal' 'stl' 'chi, phi' 'min' 't.b' 'dal, t.b' 'col' 'det, l.a' 'nyr'
 'phi' 'fla' 'buf, s.j' 'n.j, stl' 'nyr, stl' 's.j' 'atl, pit' 'cbj, edm'
 'phx' 'car' 'cbj' 'min, tor' 'pit, tor' 'cgy, chi, phi' 'car, ott'
 'col, fla' 'ana, nyi' 'van, wsh' 'cbj, wsh' 'ana, bos' 'dal, min'
 'buf, edm' 'stl, tor' 'cbj, ott' 'dal, stl' 'nyi, pit' 'ana, dal'
 'nyi, ott' 'cbj, t.b' 'ana, pit' 'min, nyr' 'phi, phx' 'dal, pit'
 'nyr, phx' 'buf, tor' 'dal, mtl' 'ana, atl' 'car, edm' 'mtl, phi'
 'ana, chi' 'atl, cbj' 'car, t.b' 'car, l.a' 'col, van' 'bos, t.b'
 'atl, mtl' 'cgy, phx' 'ana, van' 'fla, nsh' 'fla, t.b' 'edm, l.a'
 'dal, nyr' 't.b, van' 'fla, phi, t.b' 'ana, s.j' 'cgy, edm' 'ana, car'
 'cgy, nyr' 'car, van' 'ana, chi, min' 'min, mtl' 'car, wsh' 'bos, fla'
 'chi, min' 'bos, min' 'atl, buf' 'bos, buf' 'edm, nsh' 'bos, phx'
 'cgy, tor' 'fla, mtl' 'det, t.b

In [52]:
num_records = data[data['TEAM_y'] == '-'].shape[0]
print(num_records)

6858


Having 481 different values for TEAM seems unweildy, but there are 6858 records with '-' as the value for team, so let's keep the TEAM_x column and drop TEAM_y.

In [53]:
data.drop(columns=['TEAM_y'], inplace=True)

# Rename the column 'TEAM_x' to 'TEAM'
data.rename(columns={'TEAM_x': 'TEAM'}, inplace=True)
data.columns.to_list()

['POSITION',
 'PLAYER',
 'TEAM',
 'TOI',
 'GP',
 'TOI/GP',
 'GOALS/60',
 'TOTAL ASSISTS/60',
 'FIRST ASSISTS/60',
 'SECOND ASSISTS/60',
 'TOTAL POINTS/60',
 'IPP',
 'SHOTS/60',
 'SH%',
 'IXG/60',
 'ICF/60',
 'IFF/60',
 'ISCF/60',
 'IHDCF/60',
 'RUSH ATTEMPTS/60',
 'REBOUNDS CREATED/60',
 'PIM/60',
 'TOTAL PENALTIES/60',
 'MINOR/60',
 'MAJOR/60',
 'MISCONDUCT/60',
 'PENALTIES DRAWN/60',
 'GIVEAWAYS/60',
 'TAKEAWAYS/60',
 'HITS/60',
 'HITS TAKEN/60',
 'SHOTS BLOCKED/60',
 'FACEOFFS WON/60',
 'FACEOFFS LOST/60',
 'FACEOFFS %',
 'CF/60',
 'CA/60',
 'CF%',
 'FF/60',
 'FA/60',
 'FF%',
 'SF/60',
 'SA/60',
 'SF%',
 'GF/60',
 'GA/60',
 'GF%',
 'XGF/60',
 'XGA/60',
 'XGF%',
 'SCF/60',
 'SCA/60',
 'SCF%',
 'HDCF/60',
 'HDCA/60',
 'HDCF%',
 'HDGF/60',
 'HDGA/60',
 'HDGF%',
 'MDCF/60',
 'MDCA/60',
 'MDCF%',
 'MDGF/60',
 'MDGA/60',
 'MDGF%',
 'LDCF/60',
 'LDCA/60',
 'LDCF%',
 'LDGF/60',
 'LDGA/60',
 'LDGF%',
 'ON-ICE SH%',
 'ON-ICE SV%',
 'PDO',
 'OFF.\xa0ZONE STARTS/60',
 'NEU.\xa0ZONE STARTS/60',


# Now we should do the same thing with the AGE columns

In [54]:
print(data['AGE_x'].unique())
print(data['AGE_x'].nunique())
print(data['AGE_y'].unique())
print(data['AGE_y'].nunique())

[41 52 43 40 45 38 44 37 51 48 42 36 46 49 53 47 39 50 35 54 34 55 62 57
 33 32 56 31 30 29 28 27 26 25 24 23 22 21 20 19 18]
41
['24' '35' '26' '23' '28' '21' '27' '20' '34' '31' '29' '32' '25' '19'
 '36' '30' '22' '33' 'Deceased (28)' '37' '18' 'Deceased (37)' '17'
 'Deceased (34)' '38' 'Deceased (36)' '39' '46' '42' 'Deceased (38)'
 'Deceased (35)' '47' '41' '40' 'Deceased (23)' 'Deceased (31)'
 'Deceased (29)' 'Deceased (25)' '43' '44']
40


The values from our AGE_x columns unfortunately do not make a lot of sense, seeing as nobody who is 62 years old has ever played in the NHL. On the other hand,the values in our AGE_y column DO make a lot of sense, as there have been players who play and get paid up until they are in their mid forties, however it is very rare.

## How to proceed?

I think we should make a feature "Deceased" that extracts all rows where 'AGE_y' starts with 'Deceased', then convert 'AGE_y' to numeric, then replace 'AGE_x' with 'AGE_y', then finally rename 'AGE_x' to simply 'AGE'.

In [55]:
# Create a new feature 'DECEASED' and one-hot-encode it
data['DECEASED'] = np.where(data['AGE_y'].str.startswith('Deceased'), 1, 0)

# Print the updated dataframe
print(data['DECEASED'].unique())


[0 1]


In [56]:
# Extract the age from the 'AGE_y' column and convert it to an integer
data['AGE_y'] = data['AGE_y'].str.extract('(\d+)', expand=False).astype(int)

In [57]:
data['AGE_x'] = data['AGE_y']
data.drop(columns=['AGE_y'], inplace=True)
data.rename(columns={'AGE_x': 'AGE'}, inplace=True)

In [58]:
data.columns.to_list()

['POSITION',
 'PLAYER',
 'TEAM',
 'TOI',
 'GP',
 'TOI/GP',
 'GOALS/60',
 'TOTAL ASSISTS/60',
 'FIRST ASSISTS/60',
 'SECOND ASSISTS/60',
 'TOTAL POINTS/60',
 'IPP',
 'SHOTS/60',
 'SH%',
 'IXG/60',
 'ICF/60',
 'IFF/60',
 'ISCF/60',
 'IHDCF/60',
 'RUSH ATTEMPTS/60',
 'REBOUNDS CREATED/60',
 'PIM/60',
 'TOTAL PENALTIES/60',
 'MINOR/60',
 'MAJOR/60',
 'MISCONDUCT/60',
 'PENALTIES DRAWN/60',
 'GIVEAWAYS/60',
 'TAKEAWAYS/60',
 'HITS/60',
 'HITS TAKEN/60',
 'SHOTS BLOCKED/60',
 'FACEOFFS WON/60',
 'FACEOFFS LOST/60',
 'FACEOFFS %',
 'CF/60',
 'CA/60',
 'CF%',
 'FF/60',
 'FA/60',
 'FF%',
 'SF/60',
 'SA/60',
 'SF%',
 'GF/60',
 'GA/60',
 'GF%',
 'XGF/60',
 'XGA/60',
 'XGF%',
 'SCF/60',
 'SCA/60',
 'SCF%',
 'HDCF/60',
 'HDCA/60',
 'HDCF%',
 'HDGF/60',
 'HDGA/60',
 'HDGF%',
 'MDCF/60',
 'MDCA/60',
 'MDCF%',
 'MDGF/60',
 'MDGA/60',
 'MDGF%',
 'LDCF/60',
 'LDCA/60',
 'LDCF%',
 'LDGF/60',
 'LDGA/60',
 'LDGF%',
 'ON-ICE SH%',
 'ON-ICE SV%',
 'PDO',
 'OFF.\xa0ZONE STARTS/60',
 'NEU.\xa0ZONE STARTS/60',


# Making sure columns are of the proper data type
Before delving too far into the analysis, we should make sure the data types of our columns are all appropriate (categorical types are categorical, and numerical types are numerical).

In [59]:
# Select all columns that are non-numeric
non_numeric_columns = data.select_dtypes(exclude=[np.number]).columns.to_list()
non_numeric_columns

['POSITION',
 'PLAYER',
 'TEAM',
 'IPP',
 'SH%',
 'FACEOFFS %',
 'CF%',
 'FF%',
 'SF%',
 'GF%',
 'XGF%',
 'SCF%',
 'HDCF%',
 'HDGF%',
 'MDCF%',
 'MDGF%',
 'LDCF%',
 'LDGF%',
 'ON-ICE SH%',
 'ON-ICE SV%',
 'PDO',
 'OFF.\xa0ZONE START %',
 'OFF.\xa0ZONE FACEOFF %',
 'DATE OF BIRTH',
 'BIRTH CITY',
 'BIRTH STATE/PROVINCE',
 'BIRTH COUNTRY',
 'NATIONALITY',
 'DRAFT TEAM',
 'DRAFT ROUND',
 'ROUND PICK',
 'OVERALL DRAFT POSITION',
 'YEAR',
 'COUNTRY',
 'POS',
 'HANDED',
 'DRAFTED',
 'SIGNING TEAM',
 'TYPE',
 'CLAUSE',
 'EXPIRY',
 'CAP HIT',
 'AAV',
 'SALARY',
 'BASE SALARY',
 'S.BONUS',
 'P.BONUS',
 'SEASON']

We can see that there are quite a few columns here that *should* be numerical. In fact, probably the majority of them. I think it may be easiest to simply manually separate them.

In [60]:
# Select all columns from non_numeric_columns that contain '%' in the feature name
percentage_columns = [col for col in non_numeric_columns if '%' in col]

non_numeric_columns = list((set(non_numeric_columns) - set(percentage_columns)))

# Other numeric columns
numeric_columns = ['DRAFT ROUND', 'BASE SALARY', 'P.BONUS', 
                   'ROUND PICK', 'OVERALL DRAFT POSITION', 
                   'SALARY', 'PDO', 'AAV', 'IPP', 'CAP HIT', 
                   'S.BONUS']

In [61]:
# Convert all percentage columns to numeric
for col in percentage_columns:
    # Replace '-' with NaN
    data[col] = data[col].replace('-', np.nan)
    data[col] = data[col].str.rstrip('%').astype(float) / 100.0

# view all percentage columns
data[percentage_columns].head()

SH%  FACEOFFS %     CF%     FF%     SF%     GF%    XGF%    SCF%   HDCF%  \
0  0.0580      0.0265  0.3432  0.3555  0.3624  0.2632  0.2987  0.3513  0.3270   
1  0.0152         NaN  0.4200  0.4233  0.4257  0.4000  0.4192  0.4291  0.4555   
2  0.0513      0.0553  0.2782  0.2897  0.3038  0.2581  0.2506  0.3077  0.2557   
3  0.1087      0.0087  0.5586  0.5579  0.5553  0.6084  0.5505  0.5610  0.5536   
4  0.1111      0.0220  0.5903  0.5923  0.6020  0.6531  0.6163  0.5543  0.5732   

    HDGF%   MDCF%   MDGF%   LDCF%   LDGF%  ON-ICE SH%  ON-ICE SV%  \
0  0.2500  0.3679  0.2727  0.3412  0.2857      0.0660      0.8949   
1  0.4730  0.4100  0.3200  0.3985  0.2308      0.0756      0.9159   
2  0.1905  0.3458  0.3750  0.2580  0.5000      0.0556      0.9303   
3  0.5294  0.5650  0.6471  0.5733  0.8182      0.1203      0.9033   
4  0.6000  0.5424  0.5789  0.6205  0.7667      0.0960      0.9229   

   OFF. ZONE START %  OFF. ZONE FACEOFF %  
0             0.3878               0.3913  
1             0.3002               0.3253  
2             0.2794               0.2717  
3             0.6297               0.6253  
4             0.6855               0.6703

In [62]:
# Convert all columns that should be numeric to their proper data types
data['DRAFT ROUND'] = data['DRAFT ROUND'].replace('-', np.nan).astype('Int64')
data['BASE SALARY'] = data['BASE SALARY'].str.replace(',', '').str.replace('$', '').astype('Int64')
data['P.BONUS'] = data['P.BONUS'].str.replace(',', '').str.replace('$', '').astype('Int64')
data['ROUND PICK'] = data['ROUND PICK'].replace('-', np.nan).astype('Int64')
data['OVERALL DRAFT POSITION'] = data['OVERALL DRAFT POSITION'].replace('-', np.nan).astype('Int64')
data['SALARY'] = data['SALARY'].str.replace(',', '').str.replace('$', '').astype('Int64')
data['AAV'] = data['AAV'].str.replace(',', '').str.replace('$', '').astype('Int64')
data['CAP HIT'] = data['CAP HIT'].str.replace(',', '').str.replace('$', '').astype('Int64')
data['S.BONUS'] = data['S.BONUS'].str.replace(',', '').str.replace('$', '').astype('Int64')
data['IPP'] = data['IPP'].replace('-', np.nan).astype('float')
data['PDO'] = data['PDO'].replace('-', np.nan).astype('float')

/var/folders/jd/f_rvzywd5j7g5f9ckcp7wsxr0000gp/T/ipykernel_35480/4188256963.py:3: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  data['BASE SALARY'] = data['BASE SALARY'].str.replace(',', '').str.replace('$', '').astype('Int64')
/var/folders/jd/f_rvzywd5j7g5f9ckcp7wsxr0000gp/T/ipykernel_35480/4188256963.py:4: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  data['P.BONUS'] = data['P.BONUS'].str.replace(',', '').str.replace('$', '').astype('Int64')
/var/folders/jd/f_rvzywd5j7g5f9ckcp7wsxr0000gp/T/ipykernel_35480/4188256963.py:7: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not*

In [63]:
# Select all columns that are non-numeric from our data
non_numeric_columns = data.select_dtypes(exclude=[np.number]).columns.to_list()
non_numeric_columns

['POSITION',
 'PLAYER',
 'TEAM',
 'DATE OF BIRTH',
 'BIRTH CITY',
 'BIRTH STATE/PROVINCE',
 'BIRTH COUNTRY',
 'NATIONALITY',
 'DRAFT TEAM',
 'YEAR',
 'COUNTRY',
 'POS',
 'HANDED',
 'DRAFTED',
 'SIGNING TEAM',
 'TYPE',
 'CLAUSE',
 'EXPIRY',
 'SEASON']

We have column 'YEAR' and column 'SEASON'. Delete one?

In [64]:
# Check if 'YEAR' and 'SEASON' are identical
data['YEAR'].equals(data['SEASON'])

# Drop the 'YEAR' column
data.drop(columns=['YEAR'], inplace=True)

We have column 'POSITION' and column 'POS', delete one?

In [65]:
# Check if 'POSITION' and 'POS' are identical
data['POSITION'].equals(data['POS'])

# Check the values of 'POSITION' and 'POS'
print(data['POSITION'].unique())
print(data['POS'].unique())

# Since 'POS' is much more complicated in nature, drop it
data.drop(columns=['POS'], inplace=True)

['r' 'd' 'l' 'c, l' 'c' 'l, r' 'c, r' 'd, l']
['rw' 'rd' 'ld/rd' 'lw' 'lw, rw' 'ld' 'rw, lw' 'lw, rw, c' 'd' 'c' 'c, lw'
 'rw, c' 'lw, c' 'c, rw' 'c, lw, rw' 'c, rw, lw' 'lw, c, rw' 'rd, rw'
 'lw, d' 'rw, lw, c' 'rw, c, lw' 'ld/rd, lw' 'd, rw' 'rw, rd' 'ld, lw'
 'lw, ld']


# Saving to disk
Whew! That was a lot of cleaning, but our data looks like it should be in good shape moving forward.

In [66]:
# Save the merged data to a csv file
output_dir = '../../Data/entitiesResolved'
if not os.path.exists(output_dir):
    os.makedirs(output_dir)

output_file = output_dir + '/merged_data_clean.csv'
data.to_csv(output_file, index=False)